openedx/core/lib/cache_utils.py

```diff
@@ -6,40 +6,101 @@
+            their own sub-cache by, for example, calling RequestCache(namespace=NAMESPACE).clear() for their own
+            namespace.
+        arg_map_function (function: arg->string): Function to use for mapping the wrapped function's arguments to
+            strings to use in the cache key. If not provided, defaults to force_text, which converts the given
+            argument to a string.
+        request_cache_getter (function: args->RequestCache): Function that returns the RequestCache to use. If not
+            provided, defaults to edx_django_utils.cache.RequestCache.  If None, the function's return values are
+            not cached.
+    """
+    def outer_wrapper(f):
+        """
+        Outer wrapper that decorates the given function
+
+        Arguments:
+            f (func): the function to wrap
+        """
+        def inner_wrapper(*args, **kwargs):
             """
-            Wraps a method to memoize results.
+            Wrapper function to decorate with.
             """
-            request_cache = getattr(self, request_cache_attr_name, None)
+            # Check to see if we have a result in cache.  If not, invoke our wrapped
+            # function.  Cache and return the result to the caller.
+            if request_cache_getter:
+                request_cache = request_cache_getter(args)
+            else:
+                request_cache = RequestCache(namespace)
+
             if request_cache:
-                cache_key = '&'.join([hashvalue(arg) for arg in args])
-                if cache_key in request_cache.data.setdefault(func.__name__, {}):
-                    return request_cache.data[func.__name__][cache_key]
+                cache_key = _func_call_cache_key(f, arg_map_function, *args, **kwargs)
+                cached_response = request_cache.get_cached_response(cache_key)
+                if cached_response.is_found:
+                    return cached_response.value
 
-                result = func(self, *args, **kwargs)
+            result = f(*args, **kwargs)
 
-                request_cache.data[func.__name__][cache_key] = result
-                return result
-            else:
-                return func(self, *args, **kwargs)
-        return _wrapper
-    return _decorator
+            if request_cache:
+                request_cache.set(cache_key, result)
+
+            return result
+
+        return inner_wrapper
+    return outer_wrapper
+
+
+def _func_call_cache_key(func, arg_map_function, *args, **kwargs):
+    """
+    Returns a cache key based on the function's module
+    the function's name, and a stringified list of arguments
+    and a query string-style stringified list of keyword arguments.
+    """
+    arg_map_function = arg_map_function or force_text
+    converted_args = map(arg_map_function, args)
+    converted_kwargs = map(arg_map_function, reduce(list.__add__, map(list, sorted(kwargs.iteritems())), []))
```

Comment:
This is a very dense line. Could you make it something simpler to grok?

Would it simplify things if instead of calling `arg_map_function` separately on every arg, we instead had a `args_mapping_function` that takes as arguments the exact *args, **kwargs the function being decorated takes all at once, and then have it return a key from that? One nice aspect of that is that while we'd still have to declare it *args, **kwargs in the general decorator, when someone is actually implementing it, then can use the real variable names.

Reply:
FYI - This is mostly code from before.  I haven't change this particular aspect of it.  My main change was adding support for an external `arg_map_function`.

Regarding your question, I'd like for the caller to be as dumb as possible - when it comes to creating the cache key.  If we can keep more of the complexity of creating a unique cache-key at this layer, then the caller doesn't accidentally create colliding keys.  Or maybe I'm not understanding your suggestion?

I can see what I can do to clarify the code here though.